### PR TITLE
Improve/fix csc/csr argmax/argmin 

### PR DESCRIPTION
**Pull request by leofang**
Addressed issues:
1. For argmin/argmax, the precision of input floats does not matter since they can be compared just fine, so it doesn't make sense to always cast to `double` (`float64`).
2. However, the output int type is platform dependent (`int64` on Linux, `int32` on Windows), so we want to fix it properly to eliminate the Windows errors (see #4805).

The solution to fix both issues at once is to do template specializations using `RawModule`. This is both an improvement ***and*** a bug fix.

Close #4805.

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
This is ready!

File: cupyx/scipy/sparse/compressed.py [hunk 1]
```diff
@@ -1026,40 +1041,54 @@ def _arg_minor_reduce(self, ufunc, axis):
...
                      self.indptr[:len(self.indptr) - 1],
                      self.indptr[1:], cupy.int64(self.shape[0]),
                      value))
             if ufunc == cupy.argmin:
-                self._min_arg_reduction_kern(
-                    (self.shape[1],), (1,),
-                    (self.data.astype(cupy.float64), self.indices,
+                ker_name = 'min' + ker_name
+                ker = self._min_arg_reduction_mod.get_function(ker_name)
+                ker((self.shape[1],), (1,),
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py:1090]: these calls are duplicated, we can just move this one out of the if and delete the above one :)

File: cupyx/scipy/sparse/compressed.py [hunk 2]
```diff
@@ -1026,40 +1041,54 @@ def _arg_minor_reduce(self, ufunc, axis):
...
                      self.indptr[:len(self.indptr) - 1],
                      self.indptr[1:], cupy.int64(self.shape[1]),
                      value))
             if ufunc == cupy.argmin:
-                self._min_arg_reduction_kern(
-                    (self.shape[0],), (1,),
-                    (self.data.astype(cupy.float64), self.indices,
+                ker_name = 'min' + ker_name
+                ker = self._min_arg_reduction_mod.get_function(ker_name)
+                ker((self.shape[0],), (1,),
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py:1063]: dupped calls :), lets take the last one out.

File: cupyx/scipy/sparse/compressed.py [hunk 3]
```diff
@@ -175,10 +176,15 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
...
-        void max_arg_reduction(double* data, int* indices, int* x, int* y,
-                               int length, long long* z) {
+    # For _max_arg_reduction_mod and _min_arg_reduction_mod below, we pick
+    # the right template specialization according to input dtypes at runtime.
+    # The distinction in int types (T2) is important for portability in OS.
+
+    _max_arg_reduction_mod = core.RawModule(
+        code=r'''
+        template<typename T1, typename T2> __global__ void
+        max_arg_reduction(T1* data, int* indices, int* x, int* y,
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py:186]: Optional: these two kernels are essentially the same but with just a single operator changing.
Can we templatize them? even just a single python f-string template is fine :).

You can ignore this, or do it in another PR if you want!

Thanks

**Review by emcastillo — CHANGES_REQUESTED** [review]
Check comments please

File: cupyx/scipy/sparse/compressed.py (hunk 1 above)
Comment (leofang) [reply cupyx/scipy/sparse/compressed.py]: Good catch! Done in c7a03be.

File: cupyx/scipy/sparse/compressed.py (hunk 2 above)
Comment (leofang) [reply cupyx/scipy/sparse/compressed.py]: Done in c7a03be.

File: cupyx/scipy/sparse/compressed.py (hunk 3 above)
Comment (leofang) [reply cupyx/scipy/sparse/compressed.py]: Yeah I wanted to do this too but for some reason I forgot. Done in 82a6dc0.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
@emcastillo Windows is still happy 😆

File: cupyx/scipy/sparse/compressed.py [hunk 4]
```diff
@@ -1026,42 +992,46 @@ def _arg_minor_reduce(self, ufunc, axis):
...
+            ker((self.shape[0],), (1,),
+                (self.data, self.indices,
+                 self.indptr[:len(self.indptr) - 1],
+                 self.indptr[1:], cupy.int64(self.shape[1]),
+                 value))
+
+        elif axis == 0:
             # Create the vector to hold output
-            value = cupy.zeros(self.shape[1]).astype(cupy.int64)
+            # Note: it's important to set "int" here, following what SciPy
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py:1017]: Sorry that I did not notice before but these two braces are exactly the same, but one uses `shape[0]` and the other `shape[1]` 😄

Comment (leofang) [reply cupyx/scipy/sparse/compressed.py:1017]: Do you mean you wanna remove the repeated code?

Comment (emcastillo) [reply cupyx/scipy/sparse/compressed.py:1017]: I will do it in another PR. lets get this in

**Review by emcastillo — APPROVED** [review]

File: cupyx/scipy/sparse/compressed.py (hunk 4 above)
Comment (leofang) [reply cupyx/scipy/sparse/compressed.py:1017]: oh oh ok, I can do it too, just wasn't sure what you meant 😅

**Comment by leofang** [comment]
btw @emcastillo I suppose we should backport this to fix v8?
